### PR TITLE
Team Impersonation

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?

Copy could be improved - but core functionality is here.

Ability to allow team admins/owners to impersonate other users. This setting can be enabled and disabled in settings. 

https://www.loom.com/share/a11116b6456149449692d8a5c5917db2

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: apps/web/components/team/DisableTeamImpersonation.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+              <h2 className="font-cal font-bold leading-6 text-gray-900">
+                {t("user_impersonation_heading")}
+              </h2>
+              <Badge
+                className="ml-2 text-xs"
+                variant={!query.data?.disableImpersonation ? "success" : "gray"}>
+                {!query.data?.disableImpersonation ? t("enabled") : t("disabled")}
+              </Badge>
+            </div>
+            <p className="text-sm text-gray-700">{t("user_impersonation_description")}</p>
```
Comment (sean-brydon) [inline apps/web/components/team/DisableTeamImpersonation.tsx:45]: Copy to be improved here

Comment (zomars) [reply apps/web/components/team/DisableTeamImpersonation.tsx:45]: This should be a specific description for team impersonation.

File: apps/web/components/team/MemberListItem.tsx [hunk 2]
```diff
@@ -147,6 +149,16 @@ export default function MemberListItem(props: Props) {
...
                   <DropdownMenuSeparator className="h-px bg-gray-200" />
+                  <DropdownMenuItem>
+                    <Button
+                      onClick={() => setShowImpersonateModal(true)}
+                      color="minimal"
+                      StartIcon={LockClosedIcon}
+                      className="w-full flex-shrink-0 font-normal">
+                      {t("impersonate")}
+                    </Button>
+                  </DropdownMenuItem>
```
Comment (zomars) [inline apps/web/components/team/MemberListItem.tsx:160]: Can we show this conditionally if the user has enabled impersonations? Or at least make it appears disabled.

**Review by zomars — CHANGES_REQUESTED** [review]

File: apps/web/ee/lib/impersonation/ImpersonationProvider.ts [hunk 3]
```diff
@@ -8,56 +8,122 @@ const ImpersonationProvider = CredentialsProvider({
...
+                  id: parseInt(creds.teamId),
+                },
+              },
+            ],
+          },
+        },
+      },
+    });
+
+    if (sessionUserFromDb?.role !== "ADMIN" && sessionUserFromDb?.teams.length === 0) {
```
Comment (emrysal) [inline apps/web/ee/lib/impersonation/ImpersonationProvider.ts:50]: What happens when I am an admin but am not an owner of any team? Doesn't this break for non-team impersonation?

Comment (emrysal) [inline apps/web/ee/lib/impersonation/ImpersonationProvider.ts:50]: What happens when I am an admin but am not an owner of any team? Doesn't this break for non-team impersonation?

**Review by emrysal — CHANGES_REQUESTED** [review]

**Comment by emrysal** [comment]
@zomars @sean-brydon and I had a once over the ImpersonationProvider to try and improve the logic flow - have a look and approve if you're happy 👍

**Review by zomars — APPROVED** [review]
Can't find any flaws. Thank you both and nice work @emrysal @sean-brydon

**Pull request merged**